File: internal/handlers/user_handler.go
package handlers

import (
	"context"
	"net/http"
	"time"

	"chat-backend/internal/database"
	"chat-backend/internal/models"
	"chat-backend/internal/utils"

	"github.com/gin-gonic/gin"
	"go.mongodb.org/mongo-driver/bson"
	"go.mongodb.org/mongo-driver/bson/primitive"
)

type UserHandler struct {
	db *database.Database
}

func NewUserHandler(db *database.Database) *UserHandler {
	return &UserHandler{db: db}
}

func (h *UserHandler) GetMe(c *gin.Context) {
	userID, _ := c.Get("user_id")
	userIDObj := userID.(primitive.ObjectID)

	var user models.User
	err := h.db.MongoDB.Collection("users").FindOne(
		context.Background(),
		bson.M{"_id": userIDObj},
	).Decode(&user)

	if err != nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "User not found"})
		return
	}

	// Hide phone number if user wants it hidden
	if user.HidePhoneNumber {
		user.PhoneNumber = ""
	}

	c.JSON(http.StatusOK, user)
}

func (h *UserHandler) SearchByUsername(c *gin.Context) {
	username := c.Query("username")
	if username == "" {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Username required"})
		return
	}

	var user models.User
	err := h.db.MongoDB.Collection("users").FindOne(
		context.Background(),
		bson.M{"username": username},
	).Decode(&user)

	if err != nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "User not found"})
		return
	}

	// Hide phone number if user wants it hidden
	if user.HidePhoneNumber {
		user.PhoneNumber = ""
	}

	c.JSON(http.StatusOK, user)
}

func (h *UserHandler) GetDevices(c *gin.Context) {
	userID, _ := c.Get("user_id")
	userIDObj := userID.(primitive.ObjectID)

	var user models.User
	err := h.db.MongoDB.Collection("users").FindOne(
		context.Background(),
		bson.M{"_id": userIDObj},
	).Decode(&user)

	if err != nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "User not found"})
		return
	}

	c.JSON(http.StatusOK, user.ActiveDevices)
}

func (h *UserHandler) UpdateMe(c *gin.Context) {
	userID, _ := c.Get("user_id")
	userIDObj := userID.(primitive.ObjectID)

	var updateData map[string]interface{}
	if err := c.ShouldBindJSON(&updateData); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	updateData["updated_at"] = time.Now()
	_, err := h.db.MongoDB.Collection("users").UpdateOne(
		context.Background(),
		bson.M{"_id": userIDObj},
		bson.M{"$set": updateData},
	)

	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to update user"})
		return
	}

	c.JSON(http.StatusOK, gin.H{"message": "User updated successfully"})
}

type LocationUpdate struct {
	Latitude  float64 `json:"latitude" binding:"required"`
	Longitude float64 `json:"longitude" binding:"required"`
	Address   string  `json:"address"`
}

func (h *UserHandler) UpdateLocation(c *gin.Context) {
	userID, _ := c.Get("user_id")
	userIDObj := userID.(primitive.ObjectID)

	var loc LocationUpdate
	if err := c.ShouldBindJSON(&loc); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	location := models.Location{
		Latitude:  loc.Latitude,
		Longitude: loc.Longitude,
		Address:   loc.Address,
	}

	_, err := h.db.MongoDB.Collection("users").UpdateOne(
		context.Background(),
		bson.M{"_id": userIDObj},
		bson.M{"$set": bson.M{
			"location":  location,
			"updated_at": time.Now(),
		}},
	)

	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to update location"})
		return
	}

	c.JSON(http.StatusOK, gin.H{"message": "Location updated successfully"})
}

func (h *UserHandler) GetNearbyUsers(c *gin.Context) {
	userID, _ := c.Get("user_id")
	userIDObj := userID.(primitive.ObjectID)

	// Get current user's location
	var user models.User
	err := h.db.MongoDB.Collection("users").FindOne(
		context.Background(),
		bson.M{"_id": userIDObj},
	).Decode(&user)

	if err != nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "User not found"})
		return
	}

	// Get radius from query (default 10km)
	radius := 10.0
	if r := c.Query("radius"); r != "" {
		// Parse radius if provided
	}

	// Get all users with locations
	cursor, err := h.db.MongoDB.Collection("users").Find(
		context.Background(),
		bson.M{
			"_id": bson.M{"$ne": userIDObj},
			"location": bson.M{"$exists": true},
		},
	)

	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to fetch users"})
		return
	}
	defer cursor.Close(context.Background())

	var nearbyUsers []models.User
	for cursor.Next(context.Background()) {
		var u models.User
		if err := cursor.Decode(&u); err != nil {
			continue
		}

		distance := utils.CalculateDistance(
			user.Location.Latitude, user.Location.Longitude,
			u.Location.Latitude, u.Location.Longitude,
		)

		if distance <= radius {
			nearbyUsers = append(nearbyUsers, u)
		}
	}

	c.JSON(http.StatusOK, nearbyUsers)
}

File: src/handlers/userHandler.js
'use strict';

const { ObjectId } = require('mongodb');
const { getDB } = require('../database');
const { calculateDistance } = require('../utils/location');
const { hub } = require('../websocket/hub');

async function getMe(req, res) {
  try {
    const db = getDB();
    const user = await db.collection('users').findOne({ _id: new ObjectId(req.userId) });
    if (!user) return res.status(404).json({ error: 'User not found' });

    if (user.hide_phone_number) user.phone_number = '';
    return res.status(200).json(user);
  } catch (err) {
    console.error('getMe error:', err);
    return res.status(500).json({ error: 'Internal server error' });
  }
}

async function getUserByID(req, res) {
  try {
    let userIDObj;
    try { userIDObj = new ObjectId(req.params.id); } catch { return res.status(400).json({ error: 'Invalid user ID' }); }

    const db = getDB();
    const user = await db.collection('users').findOne({ _id: userIDObj });
    if (!user) return res.status(404).json({ error: 'User not found' });

    if (user.hide_phone_number) user.phone_number = '';
    return res.status(200).json(user);
  } catch (err) {
    console.error('getUserByID error:', err);
    return res.status(500).json({ error: 'Internal server error' });
  }
}

async function searchByUsername(req, res) {
  try {
    const username = req.query.q || req.query.username;
    if (!username) return res.status(400).json({ error: 'Username required' });

    const db = getDB();
    const user = await db.collection('users').findOne({ username });
    if (!user) return res.status(404).json({ error: 'User not found' });

    if (user.hide_phone_number) user.phone_number = '';
    return res.status(200).json(user);
  } catch (err) {
    console.error('searchByUsername error:', err);
    return res.status(500).json({ error: 'Internal server error' });
  }
}

async function getUserByPhoneNumber(req, res) {
  try {
    const phoneNumber = req.query.phone_number || req.query.phone;
    if (!phoneNumber) return res.status(400).json({ error: 'phone_number required' });

    const db = getDB();
    const user = await db.collection('users').findOne({ phone_number: phoneNumber });
    if (!user) return res.status(404).json({ error: 'User not found' });

    if (user.hide_phone_number) user.phone_number = '';
    return res.status(200).json(user);
  } catch (err) {
    console.error('getUserByPhoneNumber error:', err);
    return res.status(500).json({ error: 'Internal server error' });
  }
}

async function getDevices(req, res) {
  try {
    const db = getDB();
    const user = await db.collection('users').findOne({ _id: new ObjectId(req.userId) });
    if (!user) return res.status(404).json({ error: 'User not found' });

    return res.status(200).json(user.active_devices || []);
  } catch (err) {
    console.error('getDevices error:', err);
    return res.status(500).json({ error: 'Internal server error' });
  }
}

async function updateMe(req, res) {
  try {
    const updateData = req.body;
    updateData.updated_at = new Date();

    const db = getDB();
    await db.collection('users').updateOne(
      { _id: new ObjectId(req.userId) },
      { $set: updateData }
    );

    return res.status(200).json({ message: 'User updated successfully' });
  } catch (err) {
    console.error('updateMe error:', err);
    return res.status(500).json({ error: 'Internal server error' });
  }
}

async function updateLocation(req, res) {
  try {
    const { latitude, longitude, address } = req.body;
    if (latitude === undefined || longitude === undefined) {
      return res.status(400).json({ error: 'latitude and longitude are required' });
    }

    const db = getDB();
    await db.collection('users').updateOne(
      { _id: new ObjectId(req.userId) },
      {
        $set: {
          location: { latitude, longitude, address: address || '' },
          updated_at: new Date(),
        },
      }
    );

    return res.status(200).json({ message: 'Location updated successfully' });
  } catch (err) {
    console.error('updateLocation error:', err);
    return res.status(500).json({ error: 'Internal server error' });
  }
}

async function getNearbyUsers(req, res) {
  try {
    const db = getDB();
    const userIDObj = new ObjectId(req.userId);
    const currentUser = await db.collection('users').findOne({ _id: userIDObj });
    if (!currentUser) return res.status(404).json({ error: 'User not found' });

    let radius = 10.0;
    if (req.query.radius) {
      const parsed = parseFloat(req.query.radius);
      if (!isNaN(parsed) && parsed > 0) radius = parsed;
    }

    const professionQ = (req.query.profession || '').trim();
    const filter = { _id: { $ne: userIDObj }, location: { $exists: true } };
    if (professionQ) {
      filter.profession = { $regex: professionQ, $options: 'i' };
    }

    const allUsers = await db.collection('users').find(filter).toArray();
    const nearbyUsers = allUsers.filter((u) => {
      if (!u.location) return false;
      const dist = calculateDistance(
        currentUser.location.latitude, currentUser.location.longitude,
        u.location.latitude, u.location.longitude
      );
      return dist <= radius;
    });

    return res.status(200).json(nearbyUsers);
  } catch (err) {
    console.error('getNearbyUsers error:', err);
    return res.status(500).json({ error: 'Internal server error' });
  }
}

async function checkOnlineStatus(req, res) {
  try {
    const { id } = req.params;
    const isOnline = hub.isUserOnline(id);
    return res.status(200).json({ user_id: id, is_online: isOnline });
  } catch (err) {
    console.error('checkOnlineStatus error:', err);
    return res.status(500).json({ error: 'Internal server error' });
  }
}

async function getOnlineUsers(req, res) {
  try {
    const onlineIDs = hub.getOnlineUsers();
    return res.status(200).json({ online_users: onlineIDs });
  } catch (err) {
    console.error('getOnlineUsers error:', err);
    return res.status(500).json({ error: 'Internal server error' });
  }
}

module.exports = {
  getMe,
  getUserByID,
  searchByUsername,
  getUserByPhoneNumber,
  getDevices,
  updateMe,
  updateLocation,
  getNearbyUsers,
  checkOnlineStatus,
  getOnlineUsers,
};
